### DISCUSSION TITLE
2Fauth LXC: change IP/domain with Nginx Proxy Manager

### DISCUSSION THREAD

**Discussion by Ullulaki** [community-scripts/ProxmoxVE#1109] [Q&A] [answered]
I just switched to the 2Fauth-LXC and would love to change the local IP to a local domain, as I did before with the docker variant.
Before, within docker, changing the parameters APP_URL and ASSET_URL in the docker-compose.yml did the job.

Unfortunately this isn't working in the LXC (changed here: /opt/2fauth/docker/docker-compose.yml)
Do I have to change it another place?

Keep up the great work!

**Comment by Mellowlynx** [comment]
After you edit it, did you rebuild the app?
The compose file is used during the build.

You also mentioned Nginx Proxy Manager. What happens when you add the (sub-)domain there and point it to the IP and Port of the app?
Does it give an error or something?

Some apps do not care about the used hostname, hopefully also the case here.

**Reply by Ullulaki** [reply]
weeeell... how to rebuild it? I didn't know I have to do that or that I'm even capable of😇

When using NPM I get "502 Bad Gateway".
2Fauth in Docker with edited URL-parameters work flawlessly so would guess, what I'm missing is the rebuilding.

**Reply by Mellowlynx** [reply]
I had to check, as I don't use this setup.
Can you try to edit `/opt/2fauth/.env` to the correct URL's.
Restart the LXC and try again.

**Reply by Ullulaki** [reply]
still doesn't work :-(

**Comment by k1ngErr0r** [comment]
Hey, 

I also get this issue, running LXC and behind Nginx Proxy Manager - getting gateway error no matter the settings i use in the .env; 

Attempted to make some changes to nginx instance also, but no joy and waited to see if anyone else had this issue/if a fix discussed. 

Happy to attempt any suggestions and post results

**Comment by Mellowlynx** [comment, marked as answer]
OK, this app gets an official DO NOT USE from me.
It's not optimized to work on bare-metal and will require more knowledge than should be needed.
I think for the app, Docker is a better way for most users.

Anyway, I found the problem and have a fix, but you will need to change this every time you change the URL.

- Make sure the APP_URL= is set to the full URL that you will use (https://2fa.example.tld)
- Do not set ASSET_URL=, leave it commanded out.
- Set the IP of your NPM IP after `TRUSTED_PROXIES=` (ONLY for testing you can use `*` but change before you start using it)
- `cd` to the root of the app (/opt/2fauth/)
- Execute `php artisan config:clear`
- Update the 2fauth.conf file within nginx. (nano /etc/nginx/conf.d/2fauth.conf)
  make sure that `server_name` is the URL you will be using and set as APP_URL (Without the `https://` part)
- Save and restart Nginx (`systemctl restart nginx`)

Now you should be able to access 2FAuth on the URL.
You will not be able to access it through the local IP this way.

**Reply by Ullulaki** [reply]
> OK, this app gets an official DO NOT USE from me. It's not optimized to work on bare-metal and will require more knowledge than .

sorry for my late answer, thanks a lot for your effort!
and I totally agree with you on that term! I went with a small LXC with Docker & 2FA!

**Reply by tremor021** [reply]
I'm a bit confused here.
Not sure how you imagined to change default app behavior without messing with config files?
Every method of deployment has its own configuration methods, and yes, this app is optimized for bare metal, as you get instructions how to use it that way:
`https://docs.2fauth.app/getting-started/installation/self-hosted-server/`
I don't see how editing 2 variables in a config file is "harder" than editing 2 variables in a .env file.
Anyway, its resolved.

**Reply by Ullulaki** [reply]
> I'm a bit confused here. Not sure how you imagined to change default app behavior without messing with config files? Every method of deployment has its own configuration methods, and yes, this app is optimized for bare metal, as you get instructions how to use it that way: `https://docs.2fauth.app/getting-started/installation/self-hosted-server/` I don't see how editing 2 variables in a config file is "harder" than editing 2 variables in a .env file. Anyway, its resolved.

The docker variant works easy and I got it to work with my reverse proxy before and currently i doesn't have the time to debug/testing. So I just leave it as it is:)

**Reply by ghost** [reply]
> Anyway, I found the problem and have a fix, but you will need to change this every time you change the URL.
> 
>     * Make sure the APP_URL= is set to the full URL that you will use (https://2fa.example.tld)
> 
>     * Do not set ASSET_URL=, leave it commanded out.
> 
>     * Set the IP of your NPM IP after `TRUSTED_PROXIES=` (ONLY for testing you can use `*` but change before you start using it)
> 
>     * `cd` to the root of the app (/opt/2fauth/)
> 
>     * Execute `php artisan config:clear`
> 
>     * Update the 2fauth.conf file within nginx. (nano /etc/nginx/conf.d/2fauth.conf)
>       make sure that `server_name` is the URL you will be using and set as APP_URL (Without the `https://` part)
> 
>     * Save and restart Nginx (`systemctl restart nginx`)
> 
> 
> Now you should be able to access 2FAuth on the URL. You will not be able to access it through the local IP this way.

Nice, thanks!!!

I had figured out most of it, but not the last 2 bits!

**Comment by Onoitsu2** [comment]
In your NPM proxy for 2FA, may need make sure you have something like this in the advanced section

```
location / {
proxy_pass          $forward_scheme://$server:$port;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-Host $host;
proxy_set_header X-Forwarded-Server $host;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header Host $host;
# This line below is optional and may help in some proxying cases
proxy_set_header X-Forwarded-Port $server_port;
}
```

**Reply by Mellowlynx** [reply]
Yep, that is indeed also needed.

**Comment by cr4zyduck** [comment]
Hi,

I just installed 2fauth LXC today, but I can't get it working via NPM. Is it possible that the LXC container version of Helper Scripts has different files to modify?

**Reply by ichasecorals** [reply]
No. I just installed per @Mellowlynx instructions and it works.

**Comment by ghost** [comment]
I really would like the local IP access to work, which should work by default according to people who installed outside of Helper Scripts.

I found I can add 2 URL's to the 2fauth.conf file, by doing this:

`        server_name auth.example.co.uk 192.168.8.6;`

This changes the page from an nginx default page to a blank white page.

But it seems the setting holding me back is the app config, in the .env file.

If I try both URLs, they will both work seperately. I tried adding 2 on the same line, and duplicating the line, like this:

`APP_URL=https://auth.example.co.uk,http://192.168.8.6`

and

```
APP_URL=https://auth.example.co.uk
APP_URL=http://192.168.8.6
```

Sadly neither allowed both URLs to work simultaneously. Just the public URL should be enough to have both work.

Strange.
